lib/spack/spack/dependency.py

```diff
@@ -0,0 +1,133 @@
+            a tuple of all dependency types known to Spack.
+    """
+    if deptype in (None, 'all', all):
+        return all_deptypes
+
+    elif isinstance(deptype, string_types):
+        if deptype not in all_deptypes:
+            raise ValueError('Invalid dependency type: %s' % deptype)
+        return (deptype,)
+
+    elif isinstance(deptype, (tuple, list)):
+        invalid = next((d for d in deptype if d not in all_deptypes), None)
+        if invalid:
+            raise ValueError('Invalid dependency type: %s' % invalid)
+        return tuple(sorted(deptype))
+
+    return deptype
+
+
+class Dependency(object):
+    """Class representing metadata for a dependency on a package.
+
+    This class differs from ``spack.spec.DependencySpec`` because it
+    represents metadata at the ``Package`` level.
+    ``spack.spec.DependencySpec`` is a descriptor for an actual package
+    confiuguration, while ``Dependency`` is a descriptor for a package's
+    dependency *requirements*.
+
+    A dependency is a requirement for a configuration of another package
+    that satisfies a particular spec.  The dependency can have *types*,
+    which determine *how* that packge configuration is required,
+    e.g. whether it is required for building the package, whether it
+    needs to be linked to, or whether it is needed at runtime so that
+    Spack can call commands from it.
+
+    A package can also depend on another package with *patches*. This is
+    for cases where the maintainers of one package also maintain special
+    patches for their dependencies.  If one package depends on another
+    with patches, a special version of that dependency with patches
+    applied will be built for use by the dependent package.  The patches
+    are included in the new version's spec hash to differentiate it from
+    unpatched versions of the same package, so that unpatched versions of
+    the dependency package can coexist with the patched version.
+
+    """
+    def __init__(self, pkg, spec, type=None):
+        """Create a new Dependency.
+
+        Args:
+            pkg (type): Package that has this dependency
+            spec (Spec): Spec indicating dependency requirements
+            patches (list): list of patches to apply to this dependency
+            type (sequence): strings describing dependency relationship
+        """
+        assert isinstance(spec, spack.spec.Spec)
+
+        self.pkg = pkg
+        self.spec = spec.copy()
+
+        # This dict maps condition specs to lists of Patch objects, just
+        # as the patches dict on packages does.
+        self.patches = {}
+
+        if type is None:
+            self.type = set(default_deptype)
+        else:
+            self.type = set(type)
+
+    @property
+    def name(self):
+        """Get the name of the package to be patched.
+
+        This is needed so that a ``Dependency`` acts like a ``Package``
+        object when passed to a directive.
```

Comment:
I just deleted this part of the comment since I think it's unnecessary with your suggestion to just say "the name of the dependency package".

But `Dependency` objects are passed to the `patch()` directive inside `_depends_on`.